Add Goose model for Go channels.

lredlin opened the pull request:
This model has the following characteristics:
1. It is implemented as a generic struct in Go.
2. The struct stores a buffered ring queue that is used for buffered channels and 4 boolean flags that are used for unbuffered channel synchronization.
3. Each method in the struct other than Len/Cap/Close branches on whether the buffer has capacity 0 to determine whether the channel is Unbuffered or Buffered. Buffered channels use logic very similar to the concurrent blocking queue here: https://github.com/mit-pdos/gokv/blob/main/tutorial/queue/queue.go and unbuffered channels have a more subtle barrier-like implementation
4. There are TryReceive and TrySend functions that will be used by select statements and range for loops as described in the source code comments

I described the intended translation in the comments which will come in a later PR

I tested the model by copying the correctness based tests from https://go.dev/src/runtime/chanbarrier_test.go and https://go.dev/src/runtime/chan_test.go. These all passed except for TestSelfSelect. I documented the known edge cases in select statements that aren't quite modeled correctly.

This PR will depend on support for struct generics in the Goose translator so I will get review and merge the PR for those before this but figured I'd start getting feedback on this one as well.

The unbuffered channel behavior in this model is described in the diagram below:

```mermaid
stateDiagram-v2
	[*] --> Start: Create Channel
    Start-->sender_ready: Sender arrives first
    Start-->receiver_ready: Receiver arrives first
    sender_ready-->receiver_done: Receiver completes exchange
    receiver_ready-->sender_done: Sender completes exchange
    receiver_done-->Start: Sender observes completed exchange
    sender_done-->Start: Receiver observes completed exchange
    Start-->closed
    receiver_done-->closed_and_receiver_done: Close before exchange acknowledged
    sender_done-->closed_and_sender_done: Close before exchange acknowledged
    receiver_ready-->closed
    sender_ready-->closed
    closed_and_receiver_done --> closed: Sender observes completed exchange
    closed_and_sender_done --> closed: Receiver observes completed exchange
    closed-->[*]
```

upamanyus reviewed (commented):
The channel stuff looks great. The explanations in the comments were very helpful for understanding some subtleties (e.g. the self select example).

I have some comments about the Goose generics stuff, but I'll make them on the other PR.

lredlin commented:
> The four bool flags (sender/receiver_ready/done) seem to be encoding one of 5 possible states, since, as far as I can tell, at most one of them can be set at any given time. It might be that the model code would be simpler with a 5-way enum representing these states.

Also there is actually some subtlety with close: When we are in sender_done or receiver_done we still complete the exchange even if another goroutine closes the channel in between. I added a state diagram to the PR description

zeldovich commented:
Sure, merging the current version seems like a fine plan.

zeldovich commented:
Maybe this is also for a future simplification, but: why are the `sender_ready` and `receiver_done` states needed?

If the sender arrives first and there's no receiver, it seems fine for the sender to just retry (if it's a blocking operation), or bail out (if it's a non-blocking try-send).  So, I don't think the channel needs to model a waiting sender.

Similarly, I don't see a reason for the sender to hang around waiting for the receiver to complete (the `receiver_done` state).  As far as I can tell, the current model just makes the receiver panic if the channel got closed in the meantime.  But if the channel could be closed in the middle of the receive operation, it could also have been closed before the receiver started.

So, I think I'm proposing just 3 states for an unbuffered channel: idle, receiver-waiting, or sender-done (with a value that's about to be picked up by the receiver).

lredlin commented:
> Maybe this is also for a future simplification, but: why are the `sender_ready` and `receiver_done` states needed?
> 
> If the sender arrives first and there's no receiver, it seems fine for the sender to just retry (if it's a blocking operation), or bail out (if it's a non-blocking try-send). So, I don't think the channel needs to model a waiting sender.

The non-blocking TryReceive logic needs to know if there is a sender available so they can bail if not


> Similarly, I don't see a reason for the sender to hang around waiting for the receiver to complete (the `receiver_done` state). As far as I can tell, the current model just makes the receiver panic if the channel got closed in the meantime. But if the channel could be closed in the middle of the receive operation, it could also have been closed before the receiver started.

2 things here:
1. The receiver does not panic when a channel is closed, it returns the null value along with a bool "ok". It is a common pattern to use this to indicate that no more values will be sent over a channel
2. When a sender sees that there is a waiting receiver, it still must ensure that the exchange is actually completed. This is not necessary for synchronization purposes since the sender knows the receiver has started looping in the channel receive function but if the same goroutine sends a value "v", sees that there is a waiting receiver and exits, then closes the channel before the receiver has seen "v" and the value will never be consumed by the receiver if the receiver sees "closed"(i.e. in a range for in a data pipeline) 

I initially did attempt a model with 2 bools and did not have that wait loop but discovered the concerns above painfully through proofs falling apart. I think a simpler model may be possible but the 3 state suggestion sounds a lot like my failed original plan. I'll experiment more with this and probably run any potential improved model through the gauntlet of unit tests i stole from the runtime code this time instead of painful trial and error in Coq/Iris

zeldovich commented:
Oh, yeah, good point about the TryReceive on an un-buffered channel, that indeed requires an extra `sender_ready` state to represent..

And maybe that's yet another reason why you need the `receiver_done` state, because otherwise a sender that entered the `sender_ready` state without a waiting receiver doesn't quite know when to return.  It could wait until the un-buffered channel enters an idle state, but it might also never observe an idle state, depending on how the scheduler and other channel operations are scheduled.  Strictly speaking, that's a liveness problem, but seems a bit unclean to sweep this under the liveness rug.

Thanks for the explanations!

tchajed commented:
@lredlin could you rebase and squash this onto master? It will fix up the diff in GitHub to not include the changes in #61.

tchajed reviewed (commented):
I looked at the channel model and tests - everything looks good to me. This PR far exceeds our standards so far for testing and validation of the goose model.

The only change I'd like is to add a README.md to the channel directory that has the mermaid diagram in the PR description (and some minimal explanation of why this code is here).

lredlin commented:
> I looked at the channel model and tests - everything looks good to me. This PR far exceeds our standards so far for testing and validation of the goose model.
> 
> The only change I'd like is to add a README.md to the channel directory that has the mermaid diagram in the PR description (and some minimal explanation of why this code is here).

Added the README and mermaid diagram and updated the mermaid diagram to reflect close behavior more accurately
